gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApplicationServiceImpl.java

```diff
@@ -1540,29 +1542,36 @@ private ApplicationCriteria buildSearchCriteria(ExecutionContext executionContex
         }
 
         if (applicationQuery != null) {
-            if (applicationQuery.getIds() != null && !applicationQuery.getIds().isEmpty()) {
-                criteriaBuilder.ids(applicationQuery.getIds());
+            if (CollectionUtils.isNotEmpty(applicationQuery.getGroups())) {
+                criteriaBuilder.groups(applicationQuery.getGroups());
             }
 
-            ApplicationStatus applicationStatus = null;
-            if (applicationQuery.getGroups() != null && !applicationQuery.getGroups().isEmpty()) {
-                criteriaBuilder.groups(applicationQuery.getGroups());
+            if (StringUtils.isNotBlank(applicationQuery.getName())) {
+                criteriaBuilder.name(applicationQuery.getName());
             }
 
+            ApplicationStatus applicationStatus = null;
             if (applicationQuery.getStatus() != null && !applicationQuery.getStatus().isBlank()) {
                 applicationStatus = ApplicationStatus.valueOf(applicationQuery.getStatus().toUpperCase());
                 criteriaBuilder.status(applicationStatus);
             }
 
-            if (applicationQuery.getName() != null && !applicationQuery.getName().isBlank()) {
-                criteriaBuilder.name(applicationQuery.getName());
+            if (CollectionUtils.isNotEmpty(applicationQuery.getIds())) {
+                criteriaBuilder.restrictedToIds(applicationQuery.getIds());
             }
-            if (applicationQuery.getUser() != null && !applicationQuery.getUser().isBlank()) {
+
+            if (StringUtils.isNotBlank(applicationQuery.getUser())) {
                 Set<String> userApplicationsIds = findUserApplicationsIds(executionContext, applicationQuery.getUser(), applicationStatus);
                 if (userApplicationsIds.isEmpty()) {
                     return null;
                 }
-                criteriaBuilder.ids(userApplicationsIds);
+                if (CollectionUtils.isNotEmpty(applicationQuery.getIds())) {
+                    criteriaBuilder.restrictedToIds(
+                        userApplicationsIds.stream().filter(applicationQuery.getIds()::contains).collect(toSet())
+                    );
```

Comment:
As seen together, restrictedIds collection can be empty if the user is asking for applications he can't access.